ethereumjs/ultralight
# WebRTC exploration

# WebRTC
## PR to explore WebRTC use in **Transport Layer**

### Hybrid Transport Service
Browser clients can now use a hybrid transport layer.  This adds the capacity to connect directly to other browser-clients.

- HybridTransportService
  - this.webRTC = waku / webrtc tranpsort
  - this.websocket = websocket transport

#### LibP2P / Waku
- Browser clients may discover and set up connections using the WAKU protocol.  
- In the transport layer, a Waku-LightNode "subscribes" to its NodeId.  
- Messages pushed via Waku LightPush 
  - will be encoded and decoded using the client's `NodeId`
    -  using `@waku/interfaces` Encoder/Decoder
  - Nodes initiate connections by pushing a serialized discv5 `PING` packet to the peer's `NodeId` on the WAKU `LightPush` protocol
  - Nodes subscribe to their own NodeId on `LightPush` protocol.  Discv5 packets received this way are handled using `transport.emit('packet', ...)` just like those received any other way.

1. The PortalNetwork (PING/PONG) & Discv5 authentication packets (whoareyou, challenge, handshake) are exchanged using WAKU LightPush protocol.
2. WebRTC "handshake" packets (Offer/Answer/ICE) are exchanged using WAKU LightPush protocol.
3. All further messages are exchanged directly p2p (browser-to-browser) using WebRTC protocol.

The WAKU LightPush part was built using core `@waku` and `@libp2p` packages.
The WebRTC part was built from scratch using mozilla web api (https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API)

~~#### ENR~~
~~To advertise that a node is available for RTC connections, we add 1 byte to the ENR like `ENR.set('rtc', 0x01)`.  When sending a packet, we look for this bit in a peer's ENR, and then route the outgoing packet either to an rtc datachannel, or to the websocket proxy.~~

*This part was removed as unnecessary*
Embedding a webrtc tag in the ENR does not fully provide a way to route outgoing packets in all scenarios, so this part was dropped.  The transport layer does not have access to the discv5 routing table, nor does it directly interact with ENR.  This also failed to address scenarios where we have received a packet from a peer who's ENR is not yet known (**1st PING**).

Instead, the first time the hybrid transport handles an incoming packet from a new `Multiaddr`, it will Map that multiaddr to the type of transport (websocket vs. rtc) from which it was received.   The hybrid "Send" function will check that map to determine how to send packets to that peer.  

**Without any RTC tag in the ENR...**
When we PING an unknown node we will not yet know which transport it supports.  In this case we will send the PING packet out along both transports.  When a response is received, we then know which transport (websocket or webrtc) this peer supports.  

**The PING is sent first over WAKU to prioritize webrtc connections among hybrid Ultralight nodes.**



#### Proxy / Multiaddr 
The websocket proxy has been edited to listen for an event emitted from the connecting websocket to set the port number.  This way, the udp port number can be selected during PortalNetwork.create(), instead of randomly by the Proxy.

#### CLI
Small edits made to fix some build issues.  A CLI client can now connect to a browser-client running in normal mode.

#### Browser
Small layout changes.  Header now displays address, nodeId, and ENR.
Routing table display will distinguish between RTC peers and websocket peers.  RTC connected peers will be highlighted a different color, and have the `RTC` tag in the last column.

**Issue: routing table view does not reliably display all peers**
*Two browsers may connect, but both may not appear in the display*
*They **are** connected.  The issue is in the react components*


### To test
1. **OPTIONAL** From the root directory - `npm run start-proxy`
4. From the root directory - `npm run start-browser-client`
5. Open 2 browser pages to `localhost:8080`
  -  If the PROXY is running 
    - clients should start and connect to external bootnodes, quickly building a routing table
  - If the PROXY is not running
    - clients will fail to connect to external bootnodes
7.  `COPY ENR`  from one browser, and paste into `connect to peer` input on other page.

**clients should connect over waku, perform discv5 handshake, perform webrtc handshake, and add eachother to their routing tables**
**You can now test using the Peer-Tools or Get-Block-By functions in the browser**
